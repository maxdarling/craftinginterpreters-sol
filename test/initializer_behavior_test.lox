// Test to check initializer behavior with complex scenarios

print "=== Testing Initializer Behavior ===";

class ComplexInit {
  init(value) {
    this.value = value;
    this.computed = value * 2;
    // No explicit return - should return the instance
  }

  getValue() {
    return this.value;
  }

  getComputed() {
    return this.computed;
  }
}

var complex = ComplexInit(21);
print complex; // Should print "ComplexInit instance"
print complex.getValue(); // Should print 21
print complex.getComputed(); // Should print 42

// Test with method that returns this
class ChainInit {
  init(value) {
    this.value = value;
  }

  add(x) {
    this.value = this.value + x;
    return this;
  }

  getValue() {
    return this.value;
  }
}

var chain = ChainInit(10);
print chain.add(5).add(3).getValue(); // Should print 18

print "Initializer tests completed";
