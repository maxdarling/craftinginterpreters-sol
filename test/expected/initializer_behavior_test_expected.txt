=== Testing Initializer Behavior ===
ComplexInit instance
21
42
18
Initializer tests completed
